src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py

```diff
@@ -52,22 +65,101 @@ def validate_client_parameters(namespace):
         else:
             raise ValueError("Storage account '{}' not found.".format(n.account_name))
 
-def get_content_setting_validator(settings_class):
+def validate_source_uri(namespace):
+    usage_string = 'invalid usage: supply only one of the following argument sets:' + \
+                   '\n\t   --source-uri' + \
+                   '\n\tOR --source-container --source-blob [--source-snapshot] [--source-sas]' + \
+                   '\n\tOR --source-share --source-path [--source-sas]'
+    ns = vars(namespace)
+    validate_client_parameters(namespace) # must run first to resolve storage account
+    storage_acc = ns.get('account_name', None) or os.environ.get('AZURE_STORAGE_ACCOUNT')
+    uri = ns.get('copy_source', None)
+    container = ns.pop('source_container', None)
+    blob = ns.pop('source_blob', None)
+    sas = ns.pop('source_sas', None)
+    snapshot = ns.pop('source_snapshot', None)
+    share = ns.pop('source_share', None)
+    path = ns.pop('source_path', None)
+    if uri:
+        if any([container, blob, sas, snapshot, share, path]):
+            raise ValueError(usage_string)
+        else:
+            # simplest scenario--no further processing necessary
+            return
+
+    valid_blob_source = container and blob and not share and not path
+    valid_file_source = share and path and not container and not blob and not snapshot
+
+    if valid_blob_source:
+        blob_uri = 'https://{}.blob.core.windows.net/{}/{}{}'.format(
+            storage_acc, container, blob, '?' if snapshot or sas else '')
+        if snapshot:
+            blob_uri = '{}snapshot={}'.format(blob_uri, snapshot)
+        if sas:
+            blob_uri = '{}{}{}'.format(blob_uri, '&' if snapshot else '', sas)
+        namespace.copy_source = blob_uri
+    elif valid_file_source:
+        file_uri = 'https://{}.file.core.windows.net/{}/{}'.format(
+            storage_acc, share, path)
+        if sas:
+            file_uri = '{}?{}'.format(file_uri, sas)
+        namespace.copy_source = file_uri
+    else:
+        raise ValueError(usage_string)
```

Comment:
this block can be simplified to something like:

if not valid_blob_source and not valid_file_source:
    raise
params = []
if snapshot:
    params.append('snapshot={}'.format(snapshot))
    etc.
uri = '[https://{0}.blob.core.windows.net/{1}/{2}{3}](https://{0}.blob.core.windows.net/{1}/{2}{3})' \
    .format(storage_acc,
            container if valib_blob_source else share,
            blob or path,
            '?' if params else '',
            '&'.join(params) if params else '')